From a3172895010e1253a6070b6414c2398a1ed6d886 Mon Sep 17 00:00:00 2001
From: Alex Johnston <acjohnston@google.com>
Date: Wed, 5 Feb 2025 09:27:38 +0000
Subject: [PATCH] Remove invalid URLs in terms activity

Only HTTP and HTTPS URLs are allowed
This is required to fix a security vulnebility
where an app can be installed via a link embedded
in the TermsActivity

Bug: 368319929
Test: Manual using test app provided by the reporter
Flag: EXEMPT bug fix
(cherry picked from commit 0eb898e8737fe208d48812cfa45bad243deea55e)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:f6ffc1df5407791f3f23b535eb97663694bdf380)
Merged-In: I58a5f31812ec5e1ae5675f959201d4237c888ee3
Change-Id: I58a5f31812ec5e1ae5675f959201d4237c888ee3
---
 .../common/HtmlToSpannedParser.java                  | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/src/com/android/managedprovisioning/common/HtmlToSpannedParser.java b/src/com/android/managedprovisioning/common/HtmlToSpannedParser.java
index ca1a09cf..25eec183 100644
--- a/src/com/android/managedprovisioning/common/HtmlToSpannedParser.java
+++ b/src/com/android/managedprovisioning/common/HtmlToSpannedParser.java
@@ -26,6 +26,7 @@ import android.text.SpannableStringBuilder;
 import android.text.Spanned;
 import android.text.style.ClickableSpan;
 import android.text.style.URLSpan;
+import android.webkit.URLUtil;
 
 import com.android.managedprovisioning.preprovisioning.WebActivity;
 
@@ -64,7 +65,12 @@ public class HtmlToSpannedParser {
 
         URLSpan[] urlSpans = result.getSpans(0, result.length(), URLSpan.class);
         for (URLSpan urlSpan : urlSpans) {
-            Intent intent = mUrlIntentFactory.create(urlSpan.getURL());
+            String url = urlSpan.getURL();
+            if (!isValidURL(url)) {
+                ProvisionLogger.logw("Invalid URL provided. Only HTTP and HTTPS URLs are allowed.");
+                return null;
+            }
+            Intent intent = mUrlIntentFactory.create(url);
             if (intent != null) {
                 int spanStart = result.getSpanStart(urlSpan);
                 int spanEnd = result.getSpanEnd(urlSpan);
@@ -77,6 +83,10 @@ public class HtmlToSpannedParser {
         return result;
     }
 
+    private boolean isValidURL(String url) {
+        return URLUtil.isHttpUrl(url) || URLUtil.isHttpsUrl(url);
+    }
+
     /**
      * Allows to specify an intent to handle URLs
      */
-- 
2.49.0.1077.gc0e912fd4c-goog

